File: 0015-3sum/0015-3sum.java
import java.util.*;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();

        Arrays.sort(nums);

        int n = nums.length;

        for (int i = 0; i < n - 2; i++) {

            // Skip duplicate first numbers
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }

            // Since array is sorted, no possible sum can become 0
            if (nums[i] > 0) {
                break;
            }

            // Smallest possible sum is already greater than 0
            if (nums[i] + nums[i + 1] + nums[i + 2] > 0) {
                break;
            }

            // Largest possible sum is still less than 0
            if (nums[i] + nums[n - 2] + nums[n - 1] < 0) {
                continue;
            }

            int left = i + 1;
            int right = n - 1;

            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];

                if (sum < 0) {
                    left++;
                } else if (sum > 0) {
                    right--;
                } else {
                    result.add(Arrays.asList(
                        nums[i],
                        nums[left],
                        nums[right]
                    ));

                    int leftValue = nums[left];
                    int rightValue = nums[right];

                    // Skip all duplicate values
                    while (left < right && nums[left] == leftValue) {
                        left++;
                    }

                    while (left < right && nums[right] == rightValue) {
                        right--;
                    }
                }
            }
        }

        return result;
    }
}

File: 0015-3sum/0015-3sum.ts
var threeSum = function(nums) {
    const result = [];

    // Sort numbers in ascending order
    nums.sort((a, b) => a - b);

    const n = nums.length;

    for (let i = 0; i < n - 2; i++) {

        // Skip duplicate first numbers
        if (i > 0 && nums[i] === nums[i - 1]) {
            continue;
        }

        // Since the array is sorted, no possible triplet can sum to 0
        if (nums[i] > 0) {
            break;
        }

        let left = i + 1;
        let right = n - 1;

        while (left < right) {
            const sum = nums[i] + nums[left] + nums[right];

            if (sum === 0) {
                result.push([
                    nums[i],
                    nums[left],
                    nums[right]
                ]);

                // Move both pointers
                left++;
                right--;

                // Skip duplicates
                while (
                    left < right &&
                    nums[left] === nums[left - 1]
                ) {
                    left++;
                }

                while (
                    left < right &&
                    nums[right] === nums[right + 1]
                ) {
                    right--;
                }

            } else if (sum < 0) {
                // Need a bigger sum
                left++;
            } else {
                // Need a smaller sum
                right--;
            }
        }
    }

    return result;
};